lua/include/packet.lua

```diff
@@ -182,133 +134,251 @@ function pkt:enableTimestamps()
+function packetDump(self, bytes) 
+	bytes = bytes or ffi.sizeof(self.getName())
 
-local icmp6PacketType = ffi.typeof("struct icmp_v6_packet*")
---- Retrieve an ICMPv6 packet.
--- @return Packet in 'struct icmp_v6_packet' format
-function pkt:getIcmp6Packet()
-	return icmp6PacketType(self.pkt.data)
-end
+	-- print timestamp
+	write(getTimeMicros())
 
---- Retrieve either an ICMPv4 or ICMPv6 packet.
--- @return ipv4 If true or nil returns ICMPv4, ICMPv6 otherwise
--- @return Packet in 'struct icmp_packet' or 'struct icmp_v6_packet' format
-function pkt:getIcmpPacket(ipv4)
-	ipv4 = ipv4 == nil or ipv4
-	if ipv4 then
-		return self:getIcmp4Packet()
-	else
-		return self:getIcmp6Packet()
+	-- headers in cleartext
+	for i, v in ipairs(self:getHeaders()) do
+		local str = v:getString()
+		if i == 1 then write(" " .. str .. "\n") else print(str) end
 	end
+
+	-- hex dump
+	dumpHex(self, bytes)
 end
+	
+--- TODO copy old description
+function packetFill(self, namedArgs) 
+	-- fill headers
+	local headers = self:getHeaders()
+	local args = self:getArgs()
+	local accumulatedLength = 0
+	for i, v in ipairs(headers) do
+		local nextHeader = args[i + 1]
+		if type(nextHeader) == "table" then
+			nextHeader = nextHeader[1]
+		end
+		namedArgs = v:setDefaultNamedArgs(namedArgs, nextHeader, accumulatedLength)
+		v:fill(namedArgs) 
 
-local udpPacketType = ffi.typeof("struct udp_packet*")
---- Retrieve an IPv4 UDP packet.
--- @return Packet in 'struct udp_packet' format.
-function pkt:getUdp4Packet()
-	return udpPacketType(self.pkt.data)
+		accumulatedLength = accumulatedLength + ffi.sizeof(v)
+	end
 end
 
-local udp6PacketType = ffi.typeof("struct udp_v6_packet*")
---- Retrieve an IPv6 UDP packet.
--- @return Packet in 'struct udp_v6_packet' format.
-function pkt:getUdp6Packet()
-	return udp6PacketType(self.pkt.data)
+--- TODO copy old description
+function packetGet(self) 
+	local args = {} 
+	for _, v in ipairs(self:getHeaders()) do 
+		args = mergeTables(args, v:get()) 
+	end 
+	return args 
 end
 
---- Retrieve either an UDPv4 or UDPv6 packet.
--- @return ipv4 If true or nil returns UDPv4, UDPv6 otherwise
--- @return Packet in 'struct udp_packet' or 'struct udp_v6_packet' format
-function pkt:getUdpPacket(ipv4)
-	ipv4 = ipv4 == nil or ipv4
-	if ipv4 then
-		return self:getUdp4Packet()
+--- Try to find out what the next header in the payload of this packet is
+function packetResolveLastHeader(self)
+	local name = self.getName()
+	local headers = self:getHeaders()
+	local next_header = headers[#headers]:resolveNextHeader()
+	
+	if not next_header then
+		return self
 	else
-		return self:getUdp6Packet()
+		next_member = next_header
+		-- ugly, but necessary as those are always renamed
+		if next_header == "ethernet" then
+			next_member = "eth"
+		elseif next_header == "ip4" or next_header == "ip6" then
+			next_member = "ip"
+		end
+
+		name = name .. "__" .. next_header .. "_" .. next_member .. "*"
+		return ffi.cast(name, self):resolveLastHeader()
```

Comment:
All of this doesn't really work (only if it's a "standard" packet for which we already created the cstruct). Instead try the cast, if it doesn't exist yet create the struct. This only used for buf:dump() anyway (which doesn't have to be fast).